modules/samples/virtual_firealarm/src/org.wso2.carbon.device.mgt.iot.sample.virtual.firealarm.service.impl/src/main/java/org/wso2/carbon/device/mgt/iot/sample/virtual/firealarm/service/impl/util/xmpp/XMPPClient.java

```diff
@@ -0,0 +1,241 @@
+ * received. Makes use of the 'Smack-XMPP' library provided by jivesoftware/igniterealtime.
+ * <p/>
+ * It is an abstract class with an abstract method 'processXMPPMessage' allowing the user to have
+ * their own implementation of the actions to be taken upon receiving an appropriate XMPP message.
+ */
+public abstract class XMPPClient {
+	private static final Log log = LogFactory.getLog(XMPPClient.class);
+
+	private int replyTimeoutInterval = 500;    // millis
+	private String server;
+	private int port;
+	private ConnectionConfiguration config;
+	private XMPPConnection connection;
+	private PacketFilter filter;
+	private PacketListener listener;
+
+	/**
+	 * Constructor for XMPPClient passing server-IP and the XMPP-port.
+	 *
+	 * @param server the IP of the XMPP server.
+	 * @param port   the XMPP server's port to connect to. (default - 5222)
+	 */
+	public XMPPClient(String server, int port) {
+		this.server = server;
+		this.port = port;
+		initXMPPClient();
+	}
+
+	/**
+	 * Initializes the XMPP Client.
+	 * Sets the time-out-limit whilst waiting for XMPP-replies from server. Creates the XMPP
+	 * configurations to connect to the server and creates the XMPPConnection object used for
+	 * connecting and Logging-In.
+	 */
+	private void initXMPPClient() {
+		log.info(String.format(
+				"Initializing connection to XMPP Server at %1$s via port %2$d......", server,
+				port));
+		SmackConfiguration.setPacketReplyTimeout(replyTimeoutInterval);
+		config = new ConnectionConfiguration(server, port);
+//		TODO:: Need to enable SASL-Authentication appropriately
+		config.setSASLAuthenticationEnabled(false);
+		config.setSecurityMode(ConnectionConfiguration.SecurityMode.disabled);
+		connection = new XMPPConnection(config);
+	}
+
+	/**
+	 * Connects to the XMPP-Server and if successfully established connection, then tries to Log
+	 * in using the device's XMPP Account credentials. (The XMPP-Account specific to the device is
+	 * created in the XMPP server whilst downloading the Agent from the IoT Server) .
+	 *
+	 * @param username the username of the device's XMPP-Account.
+	 * @param password the password of the device's XMPP-Account.
+	 * @param resource the resource the resource, specific to the XMPP-Account to which the login
+	 *                 is made to
+	 * @throws DeviceManagementException in the event of 'Connecting to' or 'Logging into' the
+	 *                                     XMPP server fails.
+	 */
+	public void connectAndLogin(String username, String password, String resource)
+			throws DeviceManagementException {
+		try {
+			connection.connect();
+			log.info(String.format(
+					"Connection to XMPP Server at %1$s established successfully......", server));
+
+		} catch (XMPPException xmppExcepion) {
+			String errorMsg =
+					"Connection attempt to the XMPP Server at " + server + " via port " + port +
+							" failed.";
+			log.info(errorMsg);
```

Comment:
printing an error message as an info log would not be appropriate. better using log.error()....